umi.wdl

```diff
@@ -0,0 +1,102 @@
+version 1.0
+
+# Copyright (c) 2022 Leiden University Medical Center
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+task BamReadNameToUmiTag {
+
+    # This task processes a bam file with reads that have been extracted with
+    # umi-tools extract. The UMI is extracted from the read name again and put 
+    # in the bam file again with umiTag (default RX)
+    input {
+        File inputBam
+        String outputPath = "output.bam"
+        String umiTag = "RX"
+
+        String memory = "2G"
+        Int timeMinutes = 1 + ceil(size([inputBam], "G") * 10)
+        String dockerImage = "quay.io/biocontainers/pysam:0.17.0--py39h051187c_0"
+    }
+    String bamIndexPath = sub(select_first([outputPath]), "\.bam$", ".bai")
+    command <<<
+        python <<CODE
+        import pysam 
+        import sys
+        import os
+
+        from typing import Tuple
+
+        def split_umi_from_name(name) -> Tuple[str, str]:
+            id_and_rest = name.split(maxsplit=1)
+            if len(id_and_rest) == 1:
+                id, = id_and_rest
```

Comment:
Is this line correct?

Reply:
Yes. `id_and_rest` is a result of `name.split(maxsplit=1)`. This has two possible outcomes.
1. The name does not contain whitespace `My_illumina_read:klkd:42342:sads_AGGCTACG`
2. The name does contain whitespace: `asdasd:kalkjsal:asdjka_AGCGTAGC nanoporeftw anothercomment`

The `len` check tests for case 1. In which case `id_and_rest` will be `["My_illumina_read:klkd:42342:sads_AGGCTACG"]`. A list of one item. By doing `id, = id_and_rest`, the list is unpacked here into the single item ID. This is basic tuple unpacking for one item. 

If you feel it is more readable `id = id_and_rest[0]` accomplishes the same thing. I did it in the aforementioned way to keep symmetry with the other case where the list has two items.

Alternatively the code could be
```python
id_and_rest = name.split(maxsplit=1)
id = id_and_rest[0]  # name.split wll always give at least one item.
rest = id_and_rest[1] if len(id_and_rest) > 1 else ""
```
That is perhaps more clear.

Reply:
I keep the code here: https://git.lumc.nl/rhpvorderman/bam-umi-annotate. Including a few tests for the name part.

Reply:
I updated the code to be more simple and conventional.